### PR TITLE
[actpool] Improve actqueue efficiency

### PR DESCRIPTION
**Pull request by Liuhaai**
# Description

To enhance the throughput of the actpool module, multiple `queueworkers` are introduced to handle actions parallelly.
fix https://github.com/iotexproject/iotex-core/issues/3474

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

make test

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: state/factory/workingset.go [hunk 1]
```diff
@@ -465,7 +465,7 @@ func (ws *workingSet) pickAndRunActions(
 				// do nothing
 			case action.ErrChainID, errUnsupportWeb3Staking:
 				continue
-			case action.ErrGasLimit:
+			case action.ErrGasLimit, action.ErrInsufficientFunds, state.ErrNotEnoughBalance:
```
Comment (dustinxie) [inline state/factory/workingset.go:468]: we should not change this? this changes block generation logic

File: actpool/actqueue.go [hunk 2]
```diff
@@ -172,71 +179,44 @@ func (q *actQueue) cleanTimeout() []action.SealedEnvelope {
...
+	q.pendingNonce = nonce + 1
 }
 
 // PendingNonce returns the current pending nonce of the queue
 func (q *actQueue) PendingNonce() uint64 {
 	return q.pendingNonce
 }
 
+// ConfirmedNonce returns the current confirmed nonce of the queue
+func (q *actQueue) ConfirmedNonce() uint64 {
```
[the remark this answers is not in the document]
Comment (Liuhaai) [reply actpool/actqueue.go:208]: this change will be used in https://github.com/iotexproject/iotex-core/pull/3343

File: actpool/actqueue.go [hunk 3]
```diff
@@ -61,11 +61,12 @@ func (h *noncePriorityQueue) Pop() interface{} {
 type ActQueue interface {
 	Put(action.SealedEnvelope) error
 	FilterNonce(uint64) []action.SealedEnvelope
-	UpdateQueue(uint64) []action.SealedEnvelope
-	SetPendingNonce(uint64)
+	UpdateQueue() []action.SealedEnvelope
+	ConfirmedNonce() uint64
```
Comment (CoderZhi) [inline actpool/actqueue.go:65]: don't introduce this concept, especially for "Read"

**Comment by Liuhaai** [comment]
confirmedNonce is used in https://github.com/iotexproject/iotex-core/pull/3394/files#diff-b09c50140126cf69760c5606ef46e97741ab1eadf95ca82d8c021c35c5efede2R119

File: actpool/queueworker.go [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+	for {
+		job, more := <-worker.queue
+		if !more { // worker chan is closed
+			return
+		}
+		worker.Handle(job)
+	}
+}
+
+func (worker *queueWorker) Handle(job workerJob) {
```
[the remark this answers is not in the document]
Comment (Liuhaai) [reply actpool/queueworker.go:67]: it's not included in
https://github.com/iotexproject/iotex-core/pull/3377/commits/f6be2f5ef385490721860d44709ab373834d88b3

File: actpool/queueworker.go [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+	worker.mu.RUnlock()
+	// account state isn't cached in the actpool
+	if queue == nil {
+		confirmedState, err := accountutil.AccountState(ctx, worker.ap.sf, sender)
+		if err != nil {
+			return 0, nil, err
+		}
+		return confirmedState.PendingNonce(), confirmedState.Balance, nil
+	}
+	return queue.AccountNonce(), queue.AccountBalance(), nil
```
Comment (CoderZhi) [inline actpool/queueworker.go:129]: if the two numbers are not from the same view, will it be a problem?

File: actpool/queueworker.go [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+	}
+	return nil
+}
+
+func (worker *queueWorker) putAction(sender string, act action.SealedEnvelope, pendingNonce uint64, confirmedBalance *big.Int) error {
+	worker.mu.RLock()
+	queue := worker.accountActs[sender]
+	worker.mu.RUnlock()
+
+	if queue == nil {
```
Comment (CoderZhi) [inline actpool/queueworker.go:169]: is it possible that two actions from the same user are put into queue at the same time when the queue is nil, as a result, one of the action is actually abandoned?

File: actpool/queueworker.go [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+
+func (worker *queueWorker) addDestinationMap(act action.SealedEnvelope) {
+	worker.ap.accountDesActs.mu.Lock()
+	defer worker.ap.accountDesActs.mu.Unlock()
+	destn, _ := act.Destination()
+	actHash, _ := act.Hash()
+	if desQueue := worker.ap.accountDesActs.acts[destn]; desQueue == nil {
+		worker.ap.accountDesActs.acts[destn] = make(map[hash.Hash256]action.SealedEnvelope)
+	}
+	worker.ap.accountDesActs.acts[destn][actHash] = act
```
Comment (CoderZhi) [inline actpool/queueworker.go:201]: if this is related to ap.accountDesActs, maybe we should wrap it as a function of ap.accountDesActs.

File: actpool/queueworker.go [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+}
+
+func (worker *queueWorker) removeEmptyAccounts() {
```
Comment (CoderZhi) [inline actpool/queueworker.go:204]: this function is only called in `Handle`. you may consider to move the lock into `Handle`, which is a public function.

File: chainservice/chainservice.go [hunk 9]
```diff
@@ -76,11 +75,7 @@ func (cs *ChainService) HandleAction(ctx context.Context, actPb *iotextypes.Acti
 		return err
 	}
 	ctx = protocol.WithRegistry(ctx, cs.registry)
-	err = cs.actpool.Add(ctx, act)
-	if err != nil {
-		log.L().Debug(err.Error())
-	}
-	return err
+	return cs.actpool.Add(ctx, act)
```
Comment (CoderZhi) [inline chainservice/chainservice.go:78]: why the log logic is removed?

File: actpool/actpool.go [hunk 10]
```diff
@@ -317,17 +391,16 @@ func (ap *actPool) GetGasCapacity() uint64 {
...
-	ap.mutex.RLock()
-	defer ap.mutex.RUnlock()
-	pendingActs := ap.accountActs[caller.String()].AllActs()
-	ap.removeInvalidActs(pendingActs)
-	delete(ap.accountActs, caller.String())
+	worker := ap.worker[ap.allocatedWorker(caller)]
+	if queue := worker.GetQueue(caller); queue != nil {
+		pendingActs := queue.AllActs()
+		ap.removeInvalidActs(pendingActs)
+		worker.ResetAccount(caller)
```
[the remark this answers is not in the document]
Comment (Liuhaai) [reply actpool/actpool.go:402]: https://github.com/iotexproject/iotex-core/pull/3377/files/6b0a5f3041a99eef69eed90d32726bf22f9e0d97#diff-b09c50140126cf69760c5606ef46e97741ab1eadf95ca82d8c021c35c5efede2R280
do you mean the mutex here?

File: chainservice/chainservice.go (hunk 9 above)
Comment (Liuhaai) [reply chainservice/chainservice.go:78]: it has been logged in `handleActionMsg`

File: actpool/queueworker.go (hunk 6 above)
Comment (Liuhaai) [reply actpool/queueworker.go:169]: Two acts in the queue are handled by the same thread sequentially at https://github.com/iotexproject/iotex-core/pull/3377/files/6b0a5f3041a99eef69eed90d32726bf22f9e0d97#diff-b09c50140126cf69760c5606ef46e97741ab1eadf95ca82d8c021c35c5efede2R60 ?

File: actpool/queueworker.go (hunk 7 above)
Comment (Liuhaai) [reply actpool/queueworker.go:201]: Sure, good suggestions

File: actpool/queueworker.go (hunk 8 above)
Comment (Liuhaai) [reply actpool/queueworker.go:204]: yes, IMO creating a private func for the logic is ok.

**Review by CoderZhi — APPROVED** [review]

File: actpool/actpool.go [hunk 11]
```diff
@@ -141,9 +149,24 @@ func NewActPool(g genesis.Genesis, sf protocol.StateReader, cfg Config, opts ...
...
 	return ap, nil
 }
 
+// TODO: add start() and stop() in actpool
+// func (ap *actPool) Start() {
+// }
+
+// func (ap *actPool) Stop() {
+// }
+
```
Comment (dustinxie) [inline actpool/actpool.go:169]: remove if not needed

File: actpool/actpool.go [hunk 12]
```diff
@@ -157,115 +180,170 @@ func (ap *actPool) AddActionEnvelopeValidators(fs ...action.SealedEnvelopeValida
...
+func (ap *actPool) reset() {
+	var (
+		wg  sync.WaitGroup
+		ctx = ap.context(context.Background())
+	)
+	for i := range ap.worker {
+		wg.Add(1)
+		go func(worker *queueWorker) {
+			defer wg.Done()
+			worker.Reset(ctx)
```
Comment (dustinxie) [inline actpool/actpool.go:195]: just `wg.Done()` here?

File: actpool/actpool.go (hunk 11 above)
Comment (dustinxie) [reply actpool/actpool.go:169]: or should put `ap.worker[i].Start()` at L156 here?

File: actpool/actpool.go [hunk 13]
```diff
@@ -317,17 +386,14 @@ func (ap *actPool) GetGasCapacity() uint64 {
...
 	return ap.validate(ctx, selp)
 }
 
 func (ap *actPool) DeleteAction(caller address.Address) {
-	ap.mutex.RLock()
-	defer ap.mutex.RUnlock()
-	pendingActs := ap.accountActs[caller.String()].AllActs()
-	ap.removeInvalidActs(pendingActs)
-	delete(ap.accountActs, caller.String())
+	worker := ap.worker[ap.allocatedWorker(caller)]
```
Comment (dustinxie) [inline actpool/actpool.go:393]: `worker` is guaranteed `!= nil`?

File: actpool/actpool.go [hunk 14]
```diff
@@ -157,115 +180,170 @@ func (ap *actPool) AddActionEnvelopeValidators(fs ...action.SealedEnvelopeValida
...
+		_actpoolMtc.WithLabelValues("blacklisted").Inc()
+		return errors.Wrap(action.ErrAddress, "action source address is blacklisted")
 	}
 
-	caller := act.SenderAddress()
-	if caller == nil {
-		return action.ErrAddress
+	for _, ev := range ap.actionEnvelopeValidators {
+		span.AddEvent("ev.Validate")
+		if err := ev.Validate(ctx, *selp); err != nil {
```
Comment (dustinxie) [inline actpool/actpool.go:309]: is it possible to become `Validate(ctx, selp)`? maybe can be something to improve in future

File: dispatcher/dispatcher.go [hunk 15]
```diff
@@ -209,12 +211,12 @@ func (d *IotxDispatcher) EventAudit() map[iotexrpc.MessageType]int {
 }
 
 func (d *IotxDispatcher) actionHandler() {
+	d.wg.Done()
```
Comment (dustinxie) [inline dispatcher/dispatcher.go:214]: if move to here, `Stop()` can exit while some `actionHandlers()` go-routine are still running, which is not intended?

File: dispatcher/dispatcher.go [hunk 16]
```diff
@@ -223,12 +225,12 @@ func (d *IotxDispatcher) actionHandler() {
 
 // blockHandler is the main handler for handling all news from peers.
 func (d *IotxDispatcher) blockHandler() {
+	defer d.wg.Done()
```
Comment (dustinxie) [inline dispatcher/dispatcher.go:228]: see above

File: actpool/actpool.go (hunk 11 above)
Comment (Liuhaai) [reply actpool/actpool.go:168]: add as a todo task

File: actpool/actpool.go [hunk 17]
```diff
@@ -496,3 +489,19 @@ func (ap *actPool) allocatedWorker(senderAddr address.Address) int {
...
+	mu   sync.Mutex
+	acts map[string]map[hash.Hash256]action.SealedEnvelope
+}
+
+func (des *destinationMap) addAction(act action.SealedEnvelope) {
+	des.mu.Lock()
+	defer des.mu.Unlock()
+	destn, _ := act.Destination()
+	actHash, _ := act.Hash()
+	if desMap := des.acts[destn]; desMap == nil {
```
[the remark this answers is not in the document]
Comment (Liuhaai) [reply actpool/actpool.go:502]: good suggestion

File: actpool/actpool.go (hunk 13 above)
Comment (Liuhaai) [reply actpool/actpool.go:392]: `ap.worker` is a slice, and func `allocatedWorker` assure the worker is always available.

File: actpool/actpool.go (hunk 14 above)
Comment (Liuhaai) [reply actpool/actpool.go:308]: you could add an todo for it.

File: dispatcher/dispatcher.go (hunk 15 above)
Comment (Liuhaai) [reply dispatcher/dispatcher.go]: thx. it's a typo

**Review by dustinxie — APPROVED** [review]

**Pull request merged**